[MPS] Add searchsorted op

qqaatw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112830
* __->__ #112829

The metal kernels implemented are closely following `Bucketization.cu`.

Benchmark:
```
[----------------------------- searchsorted ----------------------------]
                                                         |  cpu   |  mps 
1 threads: --------------------------------------------------------------
      Batch size: 8; In features: 64; Sorter: True       |    44  |   530
      Batch size: 8; In features: 64; Sorter: False      |    31  |    12
      Batch size: 8; In features: 256; Sorter: True      |   131  |   520
      Batch size: 8; In features: 256; Sorter: False     |   107  |    12
      Batch size: 8; In features: 1024; Sorter: True     |   499  |   590
      Batch size: 8; In features: 1024; Sorter: False    |   398  |    12
      Batch size: 16; In features: 64; Sorter: True      |    71  |   540
      Batch size: 16; In features: 64; Sorter: False     |    57  |    12
      Batch size: 16; In features: 256; Sorter: True     |   242  |   610
      Batch size: 16; In features: 256; Sorter: False    |   200  |    12
      Batch size: 16; In features: 1024; Sorter: True    |   999  |   720
      Batch size: 16; In features: 1024; Sorter: False   |   842  |    12
      Batch size: 32; In features: 64; Sorter: True      |   124  |   509
      Batch size: 32; In features: 64; Sorter: False     |   103  |    12
      Batch size: 32; In features: 256; Sorter: True     |   477  |   650
      Batch size: 32; In features: 256; Sorter: False    |   407  |    12
      Batch size: 32; In features: 1024; Sorter: True    |  1940  |   833
      Batch size: 32; In features: 1024; Sorter: False   |  1710  |    12
      Batch size: 64; In features: 64; Sorter: True      |   231  |   590
      Batch size: 64; In features: 64; Sorter: False     |   194  |    12
      Batch size: 64; In features: 256; Sorter: True     |   937  |   710
      Batch size: 64; In features: 256; Sorter: False    |   800  |    13
      Batch size: 64; In features: 1024; Sorter: True    |  3980  |  1290
      Batch size: 64; In features: 1024; Sorter: False   |  3330  |    12
      Batch size: 128; In features: 64; Sorter: True     |   448  |   650
      Batch size: 128; In features: 64; Sorter: False    |   390  |    13
      Batch size: 128; In features: 256; Sorter: True    |  1830  |   850
      Batch size: 128; In features: 256; Sorter: False   |  1590  |    12
      Batch size: 128; In features: 1024; Sorter: True   |  7790  |  2850
      Batch size: 128; In features: 1024; Sorter: False  |  6670  |    13
```

qqaatw committed:
[MPS] Add searchsorted op

[ghstack-poisoned]

qqaatw commented:
Hi @albanD @malfet, can you take a look at this stack please?

malfet commented on aten/src/ATen/native/mps/operations/Bucketization.mm:
```diff
@@ -0,0 +1,377 @@
+
+  // for non-contiguous result tensors, we write the output to a contiguous copy so we can later copy back, maintaing
+  // the original result tensor
+  Tensor out = result;
+  if (!result.is_contiguous()) {
+    out = result.contiguous();
+  }
```
Nit (also, I wonder if `t.contiguous()` just returns reference to self if tensor is already contagious, in which case one does not even need a ternary)
```suggestion
  Tensor out = result.is_contiguous() ? result : result.contiguous();
```

qqaatw replied:
Yes it does.

https://github.com/pytorch/pytorch/blob/24b61a45c9e8eeb7b7bee23ad8173e41f1e59d54/aten/src/ATen/core/TensorBase.h#L124C2-L130

But we do this contiguous check, which is unnecessary, in a lot of places in the codebase. I think it would be nice to correct them all together in a separate PR?

malfet commented on aten/src/ATen/native/mps/operations/Bucketization.mm:
```diff
@@ -0,0 +1,377 @@
+#include <metal_stdlib>
+using namespace metal;
+
+/*----------Helpers--------*/
```
Please mention that those are mostly copy-n-paste from Bucketization.cu

malfet reviewed (approved):
Overal LGTM, but please add description (that says that it implements operator as a metal kernel following closely Bucketization.cu)
Also, would be good to add some sort of perf numbers (to show that it's faster than CPU or large enough tensors)

qqaatw commented on aten/src/ATen/native/mps/operations/Bucketization.mm:
```diff
@@ -0,0 +1,377 @@
+                             const c10::optional<Tensor>& sorter_opt,
+                             Tensor& result) {
+  // See [Note: hacky wrapper removal for optional tensor]
+  c10::MaybeOwned<Tensor> sorter_maybe_owned = at::borrow_from_optional_tensor(sorter_opt);
```
I don't have strong feelings on this, am just curious why `auto` is preferred over an explicitly defined type?

qqaatw commented on aten/src/ATen/native/mps/operations/Bucketization.mm:
```diff
@@ -0,0 +1,377 @@
+                                        const Tensor& boundaries,
+                                        const bool right,
+                                        const Tensor& sorter) {
+  int64_t numel_in = input.numel();
```
Sure

qqaatw commented on aten/src/ATen/native/mps/operations/Bucketization.mm:
```diff
@@ -0,0 +1,377 @@
+  resize_output(result, self.sizes());
+
+  // we have two inputs to set right, pre_check checks that they aren't set to opposites
+  bool is_right = (side_opt && *side_opt == "right") || right;
```
It was copied-n-pasted from `Bucketization.cu`, done.

qqaatw committed:
Update on "[MPS] Add searchsorted op"


The metal kernels implemented are closely following `Bucketization.cu`.

```
[----------------------------- searchsorted ----------------------------]
                                                         |  cpu   |  mps 
1 threads: --------------------------------------------------------------
      Batch size: 8; In features: 64; Sorter: True       |    44  |   530
      Batch size: 8; In features: 64; Sorter: False      |    31  |    12
      Batch size: 8; In features: 256; Sorter: True      |   131  |   520
      Batch size: 8; In features: 256; Sorter: False     |   107  |    12
      Batch size: 8; In features: 1024; Sorter: True     |   499  |   590
      Batch size: 8; In features: 1024; Sorter: False    |   398  |    12
      Batch size: 16; In features: 64; Sorter: True      |    71  |   540
      Batch size: 16; In features: 64; Sorter: False     |    57  |    12
      Batch size: 16; In features: 256; Sorter: True     |   242  |   610
      Batch size: 16; In features: 256; Sorter: False    |   200  |    12
      Batch size: 16; In features: 1024; Sorter: True    |   999  |   720
      Batch size: 16; In features: 1024; Sorter: False   |   842  |    12
      Batch size: 32; In features: 64; Sorter: True      |   124  |   509
      Batch size: 32; In features: 64; Sorter: False     |   103  |    12
      Batch size: 32; In features: 256; Sorter: True     |   477  |   650
      Batch size: 32; In features: 256; Sorter: False    |   407  |    12
      Batch size: 32; In features: 1024; Sorter: True    |  1940  |   833
      Batch size: 32; In features: 1024; Sorter: False   |  1710  |    12
      Batch size: 64; In features: 64; Sorter: True      |   231  |   590
      Batch size: 64; In features: 64; Sorter: False     |   194  |    12
      Batch size: 64; In features: 256; Sorter: True     |   937  |   710
      Batch size: 64; In features: 256; Sorter: False    |   800  |    13
      Batch size: 64; In features: 1024; Sorter: True    |  3980  |  1290
      Batch size: 64; In features: 1024; Sorter: False   |  3330  |    12
      Batch size: 128; In features: 64; Sorter: True     |   448  |   650
      Batch size: 128; In features: 64; Sorter: False    |   390  |    13
      Batch size: 128; In features: 256; Sorter: True    |  1830  |   850
      Batch size: 128; In features: 256; Sorter: False   |  1590  |    12
      Batch size: 128; In features: 1024; Sorter: True   |  7790  |  2850
      Batch size: 128; In features: 1024; Sorter: False  |  6670  |    13
```


[ghstack-poisoned]
